File: backend/schemas/savings.py
"""Request/response schemas for budget savings."""
from typing import Literal

from pydantic import BaseModel, ConfigDict, Field

SavingPeriod = Literal["monthly", "yearly"]


class SavingCreate(BaseModel):
    person: str = Field(min_length=1, max_length=120)
    account: str = Field(min_length=1, max_length=120)
    amount_annual: float = Field(ge=0)
    period: SavingPeriod = "yearly"
    pretax: bool


class SavingUpdate(BaseModel):
    person: str | None = Field(default=None, min_length=1, max_length=120)
    account: str | None = Field(default=None, min_length=1, max_length=120)
    amount_annual: float | None = Field(default=None, ge=0)
    period: SavingPeriod | None = None


class SavingOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    budget_id: int
    person: str
    account: str
    amount_annual: float
    period: SavingPeriod
    pretax: bool

File: frontend/src/api/savings.ts
// Savings data hooks. Types mirror backend/schemas/savings.py. One table
// backs both panels; `pretax` filters/creates the right rows. Mutations
// invalidate the matching savings list and ["budgets"] so the summary
// refreshes.
import { useMutation, useQuery, useQueryClient } from "@tanstack/react-query";
import { apiClient } from "./client";

export type SavingPeriod = "monthly" | "yearly";

export interface Saving {
  id: number;
  budget_id: number;
  person: string;
  account: string;
  amount_annual: number;
  period: SavingPeriod;
  pretax: boolean;
}

export interface SavingInput {
  person: string;
  account: string;
  amount_annual: number;
  period: SavingPeriod;
}

const savingKeys = {
  forBudget: (budgetId: number, pretax: boolean) => ["savings", budgetId, pretax] as const,
};

function useInvalidateAfterMutation(budgetId: number, pretax: boolean) {
  const queryClient = useQueryClient();
  return () => {
    queryClient.invalidateQueries({ queryKey: savingKeys.forBudget(budgetId, pretax) });
    queryClient.invalidateQueries({ queryKey: ["budgets"] });
  };
}

export function useSavings(budgetId: number, pretax: boolean) {
  return useQuery({
    queryKey: savingKeys.forBudget(budgetId, pretax),
    queryFn: () =>
      apiClient.get<Saving[]>(`/budgets/${budgetId}/savings?pretax=${pretax}`),
  });
}

export function useCreateSaving(budgetId: number, pretax: boolean) {
  const invalidate = useInvalidateAfterMutation(budgetId, pretax);
  return useMutation({
    mutationFn: (input: SavingInput) =>
      apiClient.post<Saving>(`/budgets/${budgetId}/savings`, { ...input, pretax }),
    onSuccess: invalidate,
  });
}

export function useUpdateSaving(budgetId: number, pretax: boolean) {
  const invalidate = useInvalidateAfterMutation(budgetId, pretax);
  return useMutation({
    mutationFn: ({ id, ...input }: Partial<SavingInput> & { id: number }) =>
      apiClient.patch<Saving>(`/budgets/${budgetId}/savings/${id}`, input),
    onSuccess: invalidate,
  });
}

export function useDeleteSaving(budgetId: number, pretax: boolean) {
  const invalidate = useInvalidateAfterMutation(budgetId, pretax);
  return useMutation({
    mutationFn: (id: number) => apiClient.del<void>(`/budgets/${budgetId}/savings/${id}`),
    onSuccess: invalidate,
  });
}
